File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/specs/GivenAStandardQueue/AndASecondFileThatThrowsAnExceptionOnPreparation/WhenTheQueueIsStarted.java
package com.lasthopesoftware.bluewater.client.playback.engine.preparation.specs.GivenAStandardQueue.AndASecondFileThatThrowsAnExceptionOnPreparation;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue;
import com.lasthopesoftware.bluewater.client.playback.file.PlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.FakeBufferingPlaybackHandler;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.FakePreparedPlayableFile;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenTheQueueIsStarted {

	private static final FakeBufferingPlaybackHandler expectedPlaybackHandler = new FakeBufferingPlaybackHandler();
	private static PlayableFile returnedPlaybackHandler;
	private static Throwable error;

	@BeforeClass
	public static void before() {

		final List<ServiceFile> serviceFiles =
			Stream
				.range(0, 2)
				.map(ServiceFile::new)
				.collect(Collectors.toList());

		final PlayableFilePreparationSource playbackPreparer = mock(PlayableFilePreparationSource.class);
		when(playbackPreparer.promisePreparedPlaybackFile(new ServiceFile(0), 0))
			.thenReturn(new Promise<>(new FakePreparedPlayableFile<>(new FakeBufferingPlaybackHandler())));

		when(playbackPreparer.promisePreparedPlaybackFile(new ServiceFile(1), 0))
			.thenReturn(new Promise<>(new Exception()))
			.thenReturn(new Promise<>(new FakePreparedPlayableFile<>(expectedPlaybackHandler)));

		final CompletingFileQueueProvider bufferingPlaybackQueuesProvider
			= new CompletingFileQueueProvider();

		final int startPosition = 0;

		final PreparedPlayableFileQueue queue = new PreparedPlayableFileQueue(
			() -> 2,
			playbackPreparer,
			bufferingPlaybackQueuesProvider.provideQueue(serviceFiles, startPosition));

		queue.promiseNextPreparedPlaybackFile(0)
			.eventually(p -> queue.promiseNextPreparedPlaybackFile(0))
			.then(pf -> returnedPlaybackHandler = pf.getPlayableFile())
			.excuse(e -> error = e);
	}

	@Test
	public void thenTheExpectedPlaybackHandlerIsNotReturned() {
		assertThat(returnedPlaybackHandler).isNotEqualTo(expectedPlaybackHandler).isNull();
	}

	@Test
	public void thenTheErrorIsCaught() {
		assertThat(error).isNotNull();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/GivenAStandardQueue/AndASecondFileThatThrowsAnExceptionOnPreparation/WhenTheQueueIsStarted.kt
package com.lasthopesoftware.bluewater.client.playback.engine.preparation.GivenAStandardQueue.AndASecondFileThatThrowsAnExceptionOnPreparation

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue
import com.lasthopesoftware.bluewater.client.playback.file.PlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakeBufferingPlaybackHandler
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakePreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toFuture
import com.namehillsoftware.handoff.promises.Promise
import org.assertj.core.api.Assertions
import org.joda.time.Duration
import org.junit.BeforeClass
import org.junit.Test
import org.mockito.Mockito
import java.util.concurrent.ExecutionException

class WhenTheQueueIsStarted {

	companion object {
		private val expectedPlaybackHandler = FakeBufferingPlaybackHandler()
		private var returnedPlaybackHandler: PlayableFile? = null
		private var error: Throwable? = null

		@JvmStatic
		@BeforeClass
		fun before() {
			val serviceFiles = intArrayOf(0, 1, 2).map { key -> ServiceFile(key) }

			val playbackPreparer = Mockito.mock(PlayableFilePreparationSource::class.java)
			Mockito.`when`(playbackPreparer.promisePreparedPlaybackFile(ServiceFile(0), Duration.ZERO))
				.thenReturn(Promise(FakePreparedPlayableFile(FakeBufferingPlaybackHandler())))
			Mockito.`when`(playbackPreparer.promisePreparedPlaybackFile(ServiceFile(1), Duration.ZERO))
				.thenReturn(Promise(Exception()))
				.thenReturn(Promise(FakePreparedPlayableFile(expectedPlaybackHandler)))

			val bufferingPlaybackQueuesProvider = CompletingFileQueueProvider()
			val startPosition = 0
			val queue = PreparedPlayableFileQueue(
				{ 2 },
				playbackPreparer,
				bufferingPlaybackQueuesProvider.provideQueue(serviceFiles, startPosition))

			try {
				returnedPlaybackHandler = queue.promiseNextPreparedPlaybackFile(Duration.ZERO)
					?.eventually { queue.promiseNextPreparedPlaybackFile(Duration.ZERO) }
					?.toFuture()?.get()?.playableFile
			} catch (e: ExecutionException) {
				error = e.cause
			}
		}
	}

	@Test
	fun thenTheExpectedPlaybackHandlerIsNotReturned() {
		Assertions.assertThat(returnedPlaybackHandler).isNotEqualTo(expectedPlaybackHandler).isNull()
	}

	@Test
	fun thenTheErrorIsCaught() {
		Assertions.assertThat(error).isNotNull
	}
}
